src/lib/storage/fake_device_factory.rb

```diff
@@ -0,0 +1,391 @@
+#!/usr/bin/env ruby
+#
+# encoding: utf-8
+
+# Copyright (c) [2015] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "storage"
+require "storage/abstract_device_factory.rb"
+require "storage/disk_size.rb"
+require "pp"
+
+module Yast
+  module Storage
+    #
+    # Factory class to generate faked devices in a device graph.
+    # This is typically used with a YaML file.
+    # Use the inherited load_yaml_file() to start the process.
+    #
+    class FakeDeviceFactory < AbstractDeviceFactory
+
+      # Valid toplevel products of this factory
+      VALID_TOPLEVEL  = [ "disk" ]
+
+      # Valid hierarchy within the products of this factory.
+      # This indicates the permitted children types for each parent.
+      VALID_HIERARCHY =
+        {
+          "disk"       => ["partition_table", "partitions", "file_system"],
+          "partitions" => ["partition", "free"],
+          "partition"  => ["file_system"]
+        }
+
+      # Valid parameters for each product of this factory.
+      # Sub-products are not listed here.
+      VALID_PARAM =
+        {
+          "disk"            => ["name", "size"],
+          "partition_table" => [],
+          "partitions"      => [],
+          "partition"       => ["size", "name", "type", "id", "mount_point", "label"],
+          "file_system"     => [],
+          "free"            => ["size"]
+        }
+
+
+      # Valid values (case insensitive) and mapping for partition_table types.
+      PARTITION_TABLE_TYPES =
+        {
+          "loop"   => ::Storage::PtType_PT_LOOP,
+          "msdos"  => ::Storage::PtType_MSDOS,
+          "ms-dos" => ::Storage::PtType_MSDOS,
+          "gpt"    => ::Storage::PtType_GPT,
+          "dasd"   => ::Storage::PtType_DASD,
+          "mac"    => ::Storage::PtType_MAC
+        }
+
+      # Valid values (case insensitive) and mapping for partition  types.
+      PARTITION_TYPES =
+        {
+          "primary"  => ::Storage::PRIMARY,
+          "extended" => ::Storage::EXTENDED,
+          "logical"  => ::Storage::LOGICAL
+        }
+
+      # Valid values (case insensitive) and mapping for partition IDs.
+      # The corresponding hex numbers can also be used.
+      PARTITION_IDS =
```

Comment:
I already talked to Arvin about that. libstorage-bgl already has some advanced enum types that might be able to do at least some of that work for us.

OTOH this might also serve as documentation what values are permitted for each of those parameters; it's not always trivial to figure out exactly how they are written. The new documentation I am currently writing should make that somewhat easier,  tough.